### ISSUE TITLE
Openssl 3.0.0 alpha3 fips provider: OAEP encryption fails

### ISSUE THREAD

**Issue by jaimeebrown** [openssl/openssl#11865]
Observed with Openssl 3.0.0-alpha3 NULL library context configured to use fips provider only.

OAEP encryption always fails with error during EVP_PKEY_encrypt.  (Btw decryption succeeds)

```
#include <openssl/provider.h>
#include <openssl/crypto.h>
#include <openssl/evp.h>
#include <openssl/err.h>
#include <openssl/rsa.h>


int main(int argc, char *argv[]) 
{
   EVP_PKEY *pkey = NULL;
   EVP_PKEY_CTX *gen_ctx;
   EVP_PKEY_CTX *enc_ctx;
   char *pval; 
   size_t outlen;
   char *data = "hello";
   int ret;
   char *out;

   ret = OSSL_PROVIDER_available(NULL, "default");
   printf("default provider available: %d\n", ret);
   ret = OSSL_PROVIDER_available(NULL, "fips");
   printf("fips provider available: %d\n", ret);

    printf("Generating RSA key\n");
    gen_ctx = EVP_PKEY_CTX_new_id(NID_rsaEncryption, NULL);   
    if (gen_ctx == NULL) goto end;
    if (EVP_PKEY_keygen_init(gen_ctx) <= 0) goto end;
        
    if (EVP_PKEY_keygen(gen_ctx, &pkey) <= 0) goto end;
    
    printf("... done generating RSA key\n");


    printf("Setting up for OAEP encrypt\n");

    pval = OPENSSL_malloc(3);
    if (pval == NULL) goto end;

    enc_ctx = EVP_PKEY_CTX_new(pkey, NULL);
    if (enc_ctx == NULL) goto end;

    if (EVP_PKEY_encrypt_init(enc_ctx) <= 0) goto end;

    if (EVP_PKEY_CTX_set_rsa_padding(enc_ctx, RSA_PKCS1_OAEP_PADDING) <= 0) goto end;
    
    if (EVP_PKEY_CTX_set_rsa_oaep_md(enc_ctx, EVP_sha256()) <= 0) goto end;
    
    if (EVP_PKEY_CTX_set_rsa_mgf1_md(enc_ctx, EVP_sha1()) <= 0) goto end;
    
    if (EVP_PKEY_CTX_set0_rsa_oaep_label(enc_ctx, pval, 3) <= 0) goto end;
    
    if (EVP_PKEY_encrypt(enc_ctx, NULL, &outlen, data, 5) <= 0) 
        goto end;
    
    out = OPENSSL_malloc(outlen);
    if (out == NULL) goto end;
    
    printf("... done\n. Doing the encrypt\n");
    if (EVP_PKEY_encrypt(enc_ctx, out, &outlen, data, 5) <= 0) 
        goto end;
    printf(".... done. Succeeded (%ld bytes output) \n", outlen);
    
end:
   printf ("end:\n");
   ERR_print_errors_fp(stderr);
   EVP_PKEY_free(pkey);
   EVP_PKEY_CTX_free(gen_ctx);
   EVP_PKEY_CTX_free(enc_ctx);
   return 1;
}
```
Output for fips provider:
```
default provider available: 0
fips provider available: 1
Generating RSA key
... done generating RSA key
Setting up for OAEP encrypt
... done
. Doing the encrypt
end:
80:7B:1B:81:C6:7F:00:00:error:rsa routines:(unknown function):passed a null parameter:../crypto/rsa/rsa_oaep.c:74:
```
Output (success) for default provider:
```
default provider available: 1
fips provider available: 0
Generating RSA key
... done generating RSA key
Setting up for OAEP encrypt
... done
. Doing the encrypt
.... done. Succeeded (256 bytes output) 
end:
```

**Comment by levitte** [comment]
Hmmm, so it seems that `rsa_padding_add_PKCS1_OAEP_mgf1_with_libctx()` isn't getting the MD that you passed with `EVP_PKEY_CTX_set_rsa_oaep_md()` (well, it doesn't either way, but the *name* of that MD should be passed down...

**Comment by levitte** [comment]
These lines look weird...  and wrong, as this basically means that the FIPS module generates an error unconditionally:

https://github.com/openssl/openssl/blob/5a5530a29abcf5d7ab7194d73b3807d568b06cbd/crypto/rsa/rsa_oaep.c#L70-L76

**Comment by levitte** [comment]
Those line come from afb638f137958205b6b089da8967f4775b4c9bb6

**Comment by mattcaswell** [comment]
> These lines look weird... and wrong, as this basically means that the FIPS module generates an error unconditionally:

I guess that should probably be:

````
 #ifndef FIPS_MODULE 
     if (md == NULL) 
         md = EVP_sha1(); 
 #else
     if (md == NULL) {
         RSAerr(0, ERR_R_PASSED_NULL_PARAMETER); 
         return 0;
     } 
 #endif 
````

**Comment by paulidale** [comment]
What @mattcaswell typed.

**Comment by levitte** [comment]
Fix in #11869
